UC-Berkeley-Extension/global
# Angular8 for error handling

## Description
A lot of the Angular code I found ([1](https://blog.angular-university.io/rxjs-error-handling/), [2](https://stackoverflow.com/questions/46019771/catching-errors-in-angular-httpclient), [3](http://musttoknow.com/use-interceptor-angular-error-handling-angular-authentication-using-http-client-http-interceptor/)) for error handling didn't work because there have been some significant changes to Angular from earlier versions. I installed the [Support for Angular 8](https://www.drupal.org/project/pdb/issues/3098754) patch and upgraded core dependencies to accept 9.

I got stuck trying to get an example module to work. Perhaps I don't even need it to work, but I wanted to have one example to show me the NG8 module works before doing anything else. 


* [The example Ng8 component](https://github.com/devbisht/pdb_ng8_example])

## Testing instructions
1. Pull this request
1. Run `composer install` from outside of your vm (you might need to install typings (`npm install typings --global`)
1. Build the site
    * From inside the VM, run `/var/www/html/scripts/setup.sh`
1. Manually add `core_version_requirement: ^8 || ^9` in the `pdb_ng8.info.yml` file.
1. Verify Angular 8 & Example modules are installed
1. Verify that node_modules folder exists in /web/modules/contrib/pdb/modules/pdb_ng8 and in /web/modules/custom/pdb_ng8_example
    1. If not, you might need to `npm install` on the pdb_ng8 module & maybe even in the pdb_ng8_example module
1. Open browser console and go to home
1. At some point (maybe already happened) you'll crash the site with the following exception:
    ```
    Drupal\Core\Security\UntrustedCallbackException: 
    Render #pre_render callbacks must be methods of a class that implements \Drupal\Core\Security\TrustedCallbackInterface or be an anonymous function. 
    The callback was _pdb_ng8_example_block_libraries_override. 
    See https://www.drupal.org/node/2966725 in Drupal\Core\Render\Renderer->doTrustedCallback() (line 96 of core/lib/Drupal/Core/Security/DoTrustedCallbackTrait.php).
    ```
1. Compare original [pdb_ng8_example.module](https://github.com/tboggia/pdb_ng8_example/blob/master/pdb_ng8_example.module) with my attempts at addressing the exception on this branch (/web/modules/custom/pdb_ng8_example/pdb_ng8_example.module). 
1. Note the js errors point to an angular component not being found. My understanding is that the/web/modules/custom/pdb_ng8_example/js/systemConfig.js file in the example module fixes those file paths, and that systemConfig is added to the build libraries through the example.module file.
1. Note that /web/modules/custom/pdb_ng8_example/js/systemConfig.js file has a console log at the top that is not firing. 


## Remaining tasks
